prow/spyglass/podlogartifact.go

```diff
@@ -0,0 +1,143 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/deck/jobs"
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+// PodLogArtifact holds data for reading from a specific pod log
+type PodLogArtifact struct {
+	name      string
+	buildID   string
+	podName   string
+	ja        *jobs.JobAgent
+	sizeLimit int64
+}
+
+//NewPodLogArtifact creates a new PodLogArtifact
+func NewPodLogArtifact(jobName string, buildID string, podName string, ja *jobs.JobAgent) *PodLogArtifact {
+	if jobName == "" {
+		logrus.Error("Must specify non-empty jobName")
+	}
+	if buildID == "" {
+		logrus.Error("Must specify non-empty buildID")
+	}
+	return &PodLogArtifact{
+		name:      jobName,
+		buildID:   buildID,
+		podName:   podName,
+		ja:        ja,
+		sizeLimit: 500e6,
+	}
+}
+
+// CanonicalLink returns a link to where pod logs are streamed
+func (a *PodLogArtifact) CanonicalLink() string {
+	if a.name != "" && a.buildID != "" {
+		return fmt.Sprintf("/log?job=%s&id=%s", a.name, a.buildID)
+	} else {
+		//return fmt.Sprintf("/log?podname=%s", a.podName)
+		logrus.Error("podname log endpoint unsupported, use job name and build ID")
+		return ""
+	}
+}
+
+// JobPath gets the path within the job for the pod log. Always returns build-log.txt.
+// This is because the pod log becomes the build log after the job artifact uploads
+// are complete, which should be used instead of the pod log.
+func (a *PodLogArtifact) JobPath() string {
+	return "build-log.txt"
+}
+
+// ReadAt implements reading a range of bytes from the pod logs endpoint
+func (a *PodLogArtifact) ReadAt(p []byte, off int64) (n int, err error) {
+	logs := a.fetchLogs()
+	return bytes.NewReader(logs).ReadAt(p, off)
+}
+
+// ReadAll reads all available pod logs, failing if they are too large
+func (a *PodLogArtifact) ReadAll() ([]byte, error) {
+	if a.Size() > a.sizeLimit {
+		return []byte{}, viewers.ErrFileTooLarge
+	}
+	return a.fetchLogs(), nil
+}
+
+// ReadAtMost reads at most n bytes
+func (a *PodLogArtifact) ReadAtMost(n int64) ([]byte, error) {
+	logs := a.fetchLogs()
+	reader := bytes.NewReader(logs)
+	var byteCount int64
+	var p []byte
+	for byteCount <= n {
+		b, err := reader.ReadByte()
+		if err != nil {
+			if err == io.EOF {
+				return p, nil
+			}
+			return p, err
+		}
+		p = append(p, b)
+	}
+	return p, nil
+}
+
+// ReadTail reads the last n bytes of the pod log
+func (a *PodLogArtifact) ReadTail(n int64) ([]byte, error) {
+	logs := a.fetchLogs()
```

Comment:
tail is supported in the pod log api so instead of fetching the whole log and offestting client-side, you can grab the exact size you need directly from the server.

Reply:
It doesn't seem like that behavior is supported by `JobAgent` or `kube.Client`. I'm going to make a separate PR for implementing a function equivalent to `kubectl logs pod --tail -1 --limit-bytes n` in `kube.Client`

Reply:
#8850 @kargakis